pkg/controllers/provisioningv2/rke2/machineprovision/controller.go

```diff
@@ -220,16 +252,116 @@ func (h *handler) OnRemove(_ string, obj runtime.Object) (runtime.Object, error)
+	if val, ok := d.Map("metadata", "labels")["rke.cattle.io/etcd-role"]; ok {
+		if val.(string) == "true" {
+			// we need to block removal until our the v1 node that corresponds has been removed
+			clusterName := d.Map("metadata", "labels")[CapiMachineLabel].(string)
+			if clusterName == "" {
+				return obj, fmt.Errorf("error retrieving the clustername for this etcd node, label does not appear to exist %s for dynamic machine %s", CapiMachineLabel, key)
+			}
+			cluster, err := h.rancherClusterCache.Get(d.String("metadata", "namespace"), clusterName)
+			if apierror.IsNotFound(err) {
+				// we can't find the corresponding cluster, so we can go ahead and simply allow removal of the node.
+				logrus.Debugf("[MachineProvision] Proceeding with removal of dynamic machine %s as cluster was not found.", key)
+				return h.doRemove(obj)
+			} else if err != nil {
+				return obj, err
+			}
+			if !cluster.DeletionTimestamp.IsZero() {
+				// If the cluster deletion timestamp has been set, we can blindly proceed with delete.
+				logrus.Debugf("[MachineProvision] Cluster (%s) deletion timestamp was not zero that owns dynamic machine %s. Proceeding with delete", clusterName, key)
+				return h.doRemove(obj)
+			}
+			restConfig, err := h.kubeconfigManager.GetRESTConfig(cluster, cluster.Status)
+			if err != nil {
+				return obj, err
+			}
+
+			clientset, err := kubernetes.NewForConfig(restConfig)
+			if err != nil {
+				return obj, err
+			}
+
+			logrus.Debugf("[MachineProvision] Built K8s clientset to cluster: %s while processing dynamic machine removal %s", clusterName, key)
+
+			removeAnnotation := "etcd." + v2provruntime.GetRuntimeCommand(cluster.Spec.KubernetesVersion) + ".cattle.io/remove"
+			removedNodeNameAnnotation := "etcd." + v2provruntime.GetRuntimeCommand(cluster.Spec.KubernetesVersion) + ".cattle.io/removed-node-name"
+
+			machine, err := h.getMachine(obj)
+
+			if err != nil {
+				return obj, err
+			}
+
+			if machine.Status.NodeRef == nil {
+				// Machine noderef is nil, we should just allow deletion.
+				logrus.Debugf("[MachineProvision] There was no associated K8s node with this etcd dynamicmachine %s. proceeding with deletion", key)
+				return h.doRemove(obj)
+			}
+
+			logrus.Debugf("[MachineProvision] Retrieving node %s from K8s for the dynamic machine %s", machine.Status.NodeRef.Name, key)
+
+			node, err := clientset.CoreV1().Nodes().Get(context.TODO(), machine.Status.NodeRef.Name, metav1.GetOptions{})
+			if err != nil {
+				if apierror.IsNotFound(err) {
+					logrus.Debugf("[MachineProvision] Node %s for dynamic machine %s was not found. proceeding with deletion", machine.Status.NodeRef.Name, key)
+					return h.doRemove(obj)
+				}
+				return obj, err
+			}
+
+			if val, ok := node.Annotations[removeAnnotation]; ok {
+				// check val to see if it's true, if not, continue
+				if val == "true" {
```

Comment:
One fewer nested blocks.
```suggestion
			if val, ok := node.Annotations[removeAnnotation]; ok && val == "true" {
```